src/activity_actor.cpp

```diff
@@ -1292,6 +1293,234 @@ std::unique_ptr<activity_actor> try_sleep_activity_actor::deserialize( JsonIn &j
+    workout_query.query();
+    switch( workout_query.ret ) {
+        case UILIST_CANCEL:
+            act.set_to_null();
+            act_id = activity_id::NULL_ID();
+            return;
+        case 4:
+            act_id = activity_id( "ACT_WORKOUT_HARD" );
+            intensity_modifier = 4;
+            break;
+        case 3:
+            act_id = activity_id( "ACT_WORKOUT_ACTIVE" );
+            intensity_modifier = 3;
+            break;
+        case 2:
+            act_id = activity_id( "ACT_WORKOUT_MODERATE" );
+            intensity_modifier = 2;
+            break;
+        case 1:
+        default:
+            act_id = activity_id( "ACT_WORKOUT_LIGHT" );
+            intensity_modifier = 1;
+            break;
+    }
+    int length;
+    query_int( length, _( "Train for how long (minutes): " ) );
+    if( length > 0 ) {
+        duration = length * 1_minutes;
+    } else {
+        act_id = activity_id::NULL_ID();
+        act.set_to_null();
+        return;
+    }
+    act.moves_total = to_moves<int>( duration );
+    act.moves_left = act.moves_total;
+    if( who.male ) {
+        sfx::play_activity_sound( "plmove", "fatigue_m_med", sfx::get_heard_volume( location ) );
+    } else {
+        sfx::play_activity_sound( "plmove", "fatigue_f_med", sfx::get_heard_volume( location ) );
+    }
+    who.add_msg_if_player( _( "You start your workout session." ) );
+}
+
+void workout_activity_actor::do_turn( player_activity &act, Character &who )
+{
+    if( who.get_fatigue() > fatigue_levels::DEAD_TIRED ) {
+        who.add_msg_if_player( _( "You are exhausted so you finish your workout early." ) );
+        act.set_to_null();
+        return;
+    }
+    if( who.get_thirst() > 240 ) {
+        who.add_msg_if_player( _( "You are dehydrated so you finish your workout early." ) );
+        act.set_to_null();
+        return;
+    }
+    if( !rest_mode && who.get_stamina() > who.get_stamina_max() / 3 ) {
+        who.mod_stamina( -25 - intensity_modifier );
+        if( one_in( 180 / intensity_modifier ) ) {
+            who.mod_fatigue( 1 );
+            who.mod_thirst( 1 );
+        }
+        if( calendar::once_every( 16_minutes / intensity_modifier ) ) {
+            //~ heavy breathing when excercising
+            std::string huff = _( "yourself huffing and puffing!" );
+            sounds::sound( location + tripoint_east, 2 * intensity_modifier, sounds::sound_t::speech, huff,
+                           true );
+        }
+        // morale bonus kicks in gradually after 5 minutes of exercise
+        if( calendar::once_every( 2_minutes ) &&
+            ( ( elapsed + act.moves_total - act.moves_left ) / 100 * 1_turns ) > 5_minutes ) {
+            who.add_morale( MORALE_FEELING_GOOD, intensity_modifier, 20, 6_hours, 30_minutes );
+        }
+        if( calendar::once_every( 2_minutes ) ) {
+            who.add_msg_if_player( m_debug, who.activity_level_str() );
+            who.add_msg_if_player( m_debug, act.id().c_str() );
```

Comment:
This prints the internal activity id as a message to the player. Is this intended or just for debugging?

Reply:
The `m_debug` flag makes it a debug-mode-only message. But I can drop it altogether if you think it's unnecessary.

Reply:
Right, I overlooked the `m_debug`. It's fine.